File: sdk/runanywhere-flutter/packages/runanywhere_genie/android/src/main/kotlin/ai/runanywhere/sdk/genie/GeniePlugin.kt
package ai.runanywhere.sdk.genie

import android.os.Build
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/**
 * RunAnywhere Genie Flutter Plugin - Android Implementation
 *
 * This plugin provides the native bridge for the Genie NPU backend on Android.
 * The actual LLM functionality is provided by RABackendGenie native libraries (.so files).
 */
class GeniePlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel

    companion object {
        private const val CHANNEL_NAME = "runanywhere_genie"
        private const val BACKEND_VERSION = "0.1.6"
        private const val BACKEND_NAME = "Genie"

        init {
            // Load Genie backend native libraries
            try {
                System.loadLibrary("rac_backend_genie_jni")
            } catch (e: UnsatisfiedLinkError) {
                // Library may not be available in all configurations
                android.util.Log.w("Genie", "Failed to load rac_backend_genie_jni: ${e.message}")
            }
        }
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, CHANNEL_NAME)
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${Build.VERSION.RELEASE}")
            }
            "getBackendVersion" -> {
                result.success(BACKEND_VERSION)
            }
            "getBackendName" -> {
                result.success(BACKEND_NAME)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}

File: sdk/runanywhere-flutter/packages/runanywhere_genie/ios/Classes/GeniePlugin.swift
import Flutter
import UIKit

/// RunAnywhere Genie Flutter Plugin - iOS Implementation
///
/// This is a stub plugin for the Flutter plugin system.
/// Genie NPU backend is Android/Snapdragon only - this plugin provides
/// platform channel compatibility but no actual NPU functionality on iOS.
public class GeniePlugin: NSObject, FlutterPlugin {

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "runanywhere_genie",
            binaryMessenger: registrar.messenger()
        )
        let instance = GeniePlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
        case "getBackendVersion":
            result("0.1.6")
        case "getBackendName":
            result("Genie")
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}
